network/messages.cc

```diff
@@ -0,0 +1,334 @@
+/*
+Copyright 2025 Google Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+#include "messages.h"
+
+#ifdef WIN32
+#    include <winsock2.h>
+#else
+#    include <netinet/in.h>
+#endif
+
+namespace Network
+{
+
+void WriteUint32ToBuffer(uint32_t value, Buffer& dest)
+{
+    uint32_t       net_val = htonl(value);
+    const uint8_t* p_val = reinterpret_cast<const uint8_t*>(&net_val);
+    dest.insert(dest.end(), p_val, p_val + sizeof(uint32_t));
+}
+
+void WriteStringToBuffer(const std::string& str, Buffer& dest)
+{
+    WriteUint32ToBuffer(static_cast<uint32_t>(str.length()), dest);
+    dest.insert(dest.end(), str.begin(), str.end());
+}
+
+absl::StatusOr<uint32_t> ReadUint32FromBuffer(const Buffer& src, size_t& offset)
+{
+    if (src.size() < offset + sizeof(uint32_t))
+    {
+        return absl::InvalidArgumentError("Buffer too small to read a uint32_t.");
+    }
+    uint32_t net_val;
+    std::memcpy(&net_val, src.data() + offset, sizeof(uint32_t));
+    offset += sizeof(uint32_t);
+    return ntohl(net_val);
+}
+
+absl::StatusOr<std::string> ReadStringFromBuffer(const Buffer& src, size_t& offset)
+{
+    absl::StatusOr<uint32_t> len_or = ReadUint32FromBuffer(src, offset);
+    if (!len_or.ok())
+    {
+        return len_or.status();
+    }
+
+    uint32_t len = len_or.value();
+    if (src.size() < offset + len)
+    {
+        return absl::InvalidArgumentError("Buffer too small for declared string length.");
+    }
+    std::string result(reinterpret_cast<const char*>(src.data() + offset), len);
+    offset += len;
+    return result;
+}
+
+absl::Status HandShakeMessage::Serialize(Buffer& dest) const
```

Comment:
does this need to return absl::Status at all if the return value is always OkStatus?